File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-flutter/lib/src/mail_provider_support.dart
enum MailProviderSupportStatus {
  builtin_registered,
  official_registered,
  official_unregistered,
  unknown,
}

final class MailProviderSupport {
  const MailProviderSupport({
    required this.providerKey,
    required this.status,
    required this.builtin,
    required this.official,
    required this.registered,
  });

  final String providerKey;
  final MailProviderSupportStatus status;
  final bool builtin;
  final bool official;
  final bool registered;
}

final class MailProviderSupportStateRequest {
  const MailProviderSupportStateRequest({
    required this.providerKey,
    required this.builtin,
    required this.official,
    required this.registered,
  });

  final String providerKey;
  final bool builtin;
  final bool official;
  final bool registered;
}

const List<String> MailProviderSupportStatuses = <String>[
  'builtin_registered',
  'official_registered',
  'official_unregistered',
  'unknown',
];

MailProviderSupportStatus resolveMailProviderSupportStatus(
  MailProviderSupportStateRequest request,
) {
  if (request.official && request.registered) {
    return request.builtin
        ? MailProviderSupportStatus.builtin_registered
        : MailProviderSupportStatus.official_registered;
  }

  if (request.official) {
    return MailProviderSupportStatus.official_unregistered;
  }

  return MailProviderSupportStatus.unknown;
}

MailProviderSupport createMailProviderSupportState(
  MailProviderSupportStateRequest request,
) {
  return MailProviderSupport(
    providerKey: request.providerKey,
    status: resolveMailProviderSupportStatus(request),
    builtin: request.builtin,
    official: request.official,
    registered: request.registered,
  );
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-java/src/main/java/com/sdkwork/Mail/metadata/mailProviderSupport.java
package com.sdkwork.Mail.metadata;

import java.util.List;

public record MailProviderSupport(
    String providerKey,
    MailProviderSupportStatus status,
    boolean builtin,
    boolean official,
    boolean registered
) {

  public enum MailProviderSupportStatus {
    builtin_registered,
    official_registered,
    official_unregistered,
    unknown
  }

  public record MailProviderSupportStateRequest(
      String providerKey,
      boolean builtin,
      boolean official,
      boolean registered
  ) {
  }

  public static final List<String> mail_PROVIDER_SUPPORT_STATUSES = List.of(
      "builtin_registered",
      "official_registered",
      "official_unregistered",
      "unknown"
  );

  public static MailProviderSupportStatus resolveMailProviderSupportStatus(
      MailProviderSupportStateRequest request
  ) {
    if (request.official() && request.registered()) {
      return request.builtin()
          ? MailProviderSupportStatus.builtin_registered
          : MailProviderSupportStatus.official_registered;
    }

    if (request.official()) {
      return MailProviderSupportStatus.official_unregistered;
    }

    return MailProviderSupportStatus.unknown;
  }

  public static MailProviderSupport createMailProviderSupportState(
      MailProviderSupportStateRequest request
  ) {
    return new MailProviderSupport(
        request.providerKey(),
        resolveMailProviderSupportStatus(request),
        request.builtin(),
        request.official(),
        request.registered()
    );
  }
}
